### PR TITLE
Optimize socket pooling

### PR DESCRIPTION
**Pull request by arhimondr**
```
== RELEASE NOTES ==

General Changes
* Optimize connection pooling to avoid running out of sockets in certain cases
```

### CODE REVIEW

**Comment by wenleix** [comment]
The Travis failure looks relevant:

```
java.util.concurrent.BrokenBarrierException
	at java.util.concurrent.CyclicBarrier.dowait(CyclicBarrier.java:207)
	at java.util.concurrent.CyclicBarrier.await(CyclicBarrier.java:435)
	at com.facebook.presto.operator.TestPageBufferClient.testCloseDuringPendingRequest(TestPageBufferClient.java:336)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

[ERROR] testExceptionFromResponseHandler(com.facebook.presto.operator.TestPageBufferClient)  Time elapsed: 6.588 s  <<< FAILURE!
java.util.concurrent.TimeoutException
	at java.util.concurrent.CyclicBarrier.dowait(CyclicBarrier.java:257)
	at java.util.concurrent.CyclicBarrier.await(CyclicBarrier.java:435)
	at com.facebook.presto.operator.TestPageBufferClient.testExceptionFromResponseHandler(TestPageBufferClient.java:376)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

File: presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java [hunk 1]
```diff
@@ -197,23 +196,17 @@ public synchronized boolean isRunning()
...
         boolean shouldSendDelete;
-        Future<?> future;
         synchronized (this) {
             shouldSendDelete = !closed;
 
             closed = true;
 
-            future = this.future;
-
+            // do not cancel the request to avoid closing a pooled connection
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java]: Is this the correct semantic when the RpcShuffleClient is Thrift based?  More generally, should this change be implemented in the HttpClient (the future it returns won't support cancellation?)

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java]: That's the way it is implemented in Drift. Basically the future it returns is not cancelled (cancellation of the future does not cancel the request). 
It has it's own pros and cons.
Basically the pros are that the connection is never closed, the cons is that if the request is large - it is never cancelled, and in theory the request / response can be in megabytes.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java]: I'm thinking about a compromise. The `Future#cancel()` method takes a boolean parameter, the `mayInterruptIfRunning`. Currently this flag is unused, and no matter what's the value the request is always cancelled. What do you think about changing this behavior to "cancel only if the mayInterruptIfRunning flag is set"?

e.g.: https://github.com/prestodb/airlift/blob/master/http-client/src/main/java/com/facebook/airlift/http/client/jetty/JettyResponseFuture.java#L59

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java]: When it's becoming even more interesting - is when the HTTP2 comes into play. With HTTP/2 it is possible to terminate the request without closing the connection. Thus the decision whether to terminate or not to terminate depends on the underlying protocol as well. 

But anyhow, I don't think there's a real downside for Presto specifically in not terminating requests. The requests are small, and the termination is inherently racy. Thus I would argue it is fine to not to terminate the requests for our usecase.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java]: @tdcmeehan @wenleix 

How about next proposal:

1. I change the implementation of the `Future#cancel()` in Airlift so when `mayInterruptIfRunning` set to false - the request won't be terminated.

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java]: I think it's reasonable.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java]: https://github.com/prestodb/airlift/pull/16

**Comment by arhimondr** [comment]
@wenleix @tdcmeehan comments addressed

**Pull request merged**